Add debug option to PDF generation script

Mister00X opened the pull request:
This commit adds a debug switch to the PDF generation script. When
activated it generates a LaTeX source code file instead of an PDF.

The debug output is enabled by adding `--tex-debug` after the language
code.

Example:
```
./10-generate-pdf-manual.sh en --tex-debug
```


My bash knowledge is very limited please have a look at this. 
(I may have missed something obvious)

Mister00X committed:
Add debug option to PDF generation script

This commit adds a debug switch to the PDF generation script. When
activated it generates a LaTeX source code file instead of an PDF.

The debug output is enabled by adding --tex-debug after the language
code.

Example:
    ./10-generate-pdf-manual.sh en --tex-debug

Mister00X commented:
Just one thing I am not sure about with my way of doing this:
Is it possible that the `||` in https://github.com/siduction/sidu-manual/blob/37941dc3a71e089982668b5800ba25d581012b7c/development/00-generate-manual.pl#L54 gets mistakenly eaten as the second argument `$2` of `10-generate-pdf-manual.sh`?

ak-li commented:
@Mister00X 
Die Idee ist gut. Zu berücksichtigen ist, dass für die Erstellung aller PDF-Handbücher und  Webseiten der Aufruf über "00-generate-manual.pl" erfolgt. Dabei wird nur ein Ländercode an 10-... übergeben.

Es muss noch

1. unter "Usage" ein Hinweis auf die Debug Funktion bei direktem Aufruf,
2. eine Prüfung auf ein zweites Argument,
3. eine Prüfung auf das richtige zweite Argument mit entsprechender Fehlermeldung 

eingebaut werden.

ak-li commented:
@Mister00X und wir warten mit dem merge bis issue 49 erledigt ist.

Commit:
Add check for existence of second argument

Add check for existence of second argument to 10-generate-manual.sh.
Also add an error message if the wrong second argument is given.

Commit:
Describe debug option in Usage.

Commit:
Fix typo

Fix typo encountered while adding debug option.

Mister00X commented:
> Zu berücksichtigen ist, dass für die Erstellung aller PDF-Handbücher und Webseiten der Aufruf über "00-generate-manual.pl" erfolgt.

Muss ich diesbezüglich etwas bestimmtes beachten?

> @Mister00X und wir warten mit dem merge bis issue 49 erledigt ist.

Klar. Das eilt ja nicht

Mister00X commented:
> 2. eine Prüfung auf ein zweites Argument,
> 
> 3. eine Prüfung auf das richtige zweite Argument mit entsprechender Fehlermeldung

Das müsste beides mit 7b8478319e8fdbf0551628dbe75ef8b4c9519cbc erledigt sein, allerdings hab ich mir das zusammen gegoogelt. Ich bin mir also nicht ganz sicher ob das so passt.  
Besonders weil ich nicht weiß, ob ich wegen `00-generate-manual.pl` noch auf irgendwas achten muss.

> 1. unter "Usage" ein Hinweis auf die Debug Funktion bei direktem Aufruf,

Das müsste mit 36abfcef4f62d9dcdbd2f2dfd49bc4e03dbd5e7e erledigt sein.

ak-li commented:
> Besonders weil ich nicht weiß, ob ich wegen `00-generate-manual.pl` noch auf irgendwas achten muss.

Von der Funktion her nichts. Es geht nur um die Sicherheit. Das mit den Optionen kein Schadcode eingegeben werden kann. Das hatte ich bei dem Aufruf der 10-... , 20-... und 30-... bisher nicht auf dem Schirm. Siehe 00-... die Zeilen 40 - 45 allerdings in perl, nicht bash.

Ich mach dir demnächst einen Vorschlag und das https://github.com/siduction/sidu-manual/commit/7b8478319e8fdbf0551628dbe75ef8b4c9519cbc nehme ich gleich mit rein, sieht ja gut aus.

Mister00X commented:
> Ich mach dir demnächst einen Vorschlag und das [7b84783](https://github.com/siduction/sidu-manual/commit/7b8478319e8fdbf0551628dbe75ef8b4c9519cbc) nehme ich gleich mit rein, sieht ja gut aus.

Perfekt!

> Von der Funktion her nichts. Es geht nur um die Sicherheit. Das mit den Optionen kein Schadcode eingegeben werden kann. Das hatte ich bei dem Aufruf der 10-... , 20-... und 30-... bisher nicht auf dem Schirm.

Oh! Ist dann mein
```sh
echo "Unknown option $2" >&2
```
in Zeile 67 dann nicht *kritisch* weil darüber auch Schadcode ausgeführt werden könnte?

Ich denke da an sowas wie 
```sh
./10-generate-pdf-manual.sh en "$(rm -rf ~/*)"
```  
Ich sollte das wohl besser durch
```sh
echo "Unknown option." >&2
```
ersetzen.

ak-li commented:
> in Zeile 67 dann nicht _kritisch_ weil darüber auch Schadcode ausgeführt werden könnte?

Einfach die Ausgabeumleitung ">&2" weglassen.

Der Code gehört ganz an den Anfang. Die letzte Zeile ist in dem bisherigen Skript schon vorhanden.
Habe das ganze ausgiebig getestet.
Übrigens hast du mich mit deiner Idee erst darauf gebracht, dass in den drei Skripten keine Plausibilitätsprüfung stattfindet. Bin immer davon ausgegangen, dass wir alle lieb und nett sind und die drei sowieso nur von 00-... aufgerufen werden. Dort ist die Prüfung eingebaut. Bei 20-... und 30-... hole ich das jetzt nach.

~~~
### default output_format
output_format=pdf

### General test
if [ $# -gt 1 ]
then
    if [ $# -eq 2 ] && [ $2 = --tex-debug ]
    then
        echo "Creating LaTeX file instead of pdf."
        output_format=tex
    else
        echo "More than two options or the second option is not '--tex-debug'"
        exit 1
        # The error message immediately gives a hint to the correct option.
    fi
fi

### country specific page options
if [ "$1" = "de" ]
~~~

Mister00X commented:
> Einfach die Ausgabeumleitung ">&2" weglassen.

Alles klar! Ich, dachte mir einfach, dass ich sicherheitshalber mal Fehler nach stderr schreibe.

> Der Code gehört ganz an den Anfang. Die letzte Zeile ist in dem bisherigen Skript schon vorhanden.
> Habe das ganze ausgiebig getestet.

Alles klar dann ändere ich das so ab.
@ak-li Wie ist es dir lieber, soll ich die Änderung oben drauf committen oder wäre es dir lieber wenn ich https://github.com/siduction/sidu-manual/commit/7b8478319e8fdbf0551628dbe75ef8b4c9519cbc nachträglich ändere damit der Verlauf sauberer ist?

ak-li commented:
> > Einfach die Ausgabeumleitung ">&2" weglassen.
> 
> Alles klar! Ich, dachte mir einfach, dass ich sicherheitshalber mal Fehler nach stderr schreibe.

Na ja, das ist jetzt doppelt gemoppelt. "$2" weglassen hast du ja schon gemacht. ">&2" hat nur Sinn wenn zuvor eine Datei als stderr definiert wurde, weil stderr per default die Konsole ist. Wichtig ist den unbekannten Code nicht auszugeben, sondern nur die eigene Meldung.

> Wie ist es dir lieber, soll ich die Änderung oben drauf committen oder wäre es dir lieber wenn ich [7b84783](https://github.com/siduction/sidu-manual/commit/7b8478319e8fdbf0551628dbe75ef8b4c9519cbc) nachträglich ändere damit der Verlauf sauberer ist?

Da bin ich überfragt. So gut kenne ich mich mit Git nicht aus um zu beurteilen, wie der merge besser funktioniert.
Wir haben den Code und probieren das später einfach aus.

ak-li commented:
PS
Beim Ausprobieren der Formate und Farben ist mir aufgefallen, dass in `10-generate-pdf-manual.sh` die Optionen
"hmargin=xxx" und "vmargin=xxx" wirkungslos sind. Einzig aktivierte Option "margin=xxx" bringt eine Änderung.
Weißt du wo man die ganzen Optionen nachlesen kann?
Im Skript steht was von "--pdf-engine=xelatex". Gefunden habe ich [overleaf](https://de.overleaf.com/learn/latex/XeLaTeX), aber egal wo ich da suche, es fehlt überall die Auflistung der möglichen Werte für die Optionen.

Mister00X commented:
> Da bin ich überfragt. So gut kenne ich mich mit Git nicht aus um zu beurteilen, wie der merge besser funktioniert.

Also dem merge müsste das eigentlich egal sein, wenn ich deinen Vorschlag einfach oben drauf commite, dann haben wir halt am ende 5 commits gemerged von denen eines https://github.com/siduction/sidu-manual/commit/7b8478319e8fdbf0551628dbe75ef8b4c9519cbc quasi rückgängig macht. Wenn ich das stattdessen interaktiv rebase bleiben es 4, dann schaut der Verlauf "schöner" aus. So lange meine PR noch nicht gemerget ist kann ich das auch problemlos machen.

Wenn du meinen Branch allerdings zu dir lokal gepullt hast und ich rebase, dann könnte sich git bei dir beim nächsten `git pull` etwas aufregen, weil ich den Verlauf den du gepullt hast umgeschrieben habe und die commits jetzt andere hashes haben.


> Beim Ausprobieren der Formate und Farben ist mir aufgefallen, dass in `10-generate-pdf-manual.sh` die Optionen
> "hmargin=xxx" und "vmargin=xxx" wirkungslos sind. Einzig aktivierte Option "margin=xxx" bringt eine Änderung.
> Weißt du wo man die ganzen Optionen nachlesen kann?

Jup.   
Das müssten Optionen für `geometry` sein die Dokumentation dazu findet sich [hier](https://ctan.org/pkg/geometry) sowohl auf Englisch als auch auf Deutsch. (`hmargin` und `vmargin` stehen auf Seite 10 der deutschen Version.)  
(Ich bin erstaunt, dass `margin` funktioniert und die anderen beiden nicht.)



> Im Skript steht was von "--pdf-engine=xelatex".

Ja das macht schon Sinn hier XeLaTeX oder LuaLaTeX zu nutzen, hauptsächlich deshalb, weil man damit Systemschriftarten nutzen kann anstatt nur TeX, METAFONT und Schriftarten die mit texlive ausgeliefert werden.  

Wenn ich mich recht erinnere gab es zuerst latex und pdflatex, dann wurde irgendwann für OS X xelatex entwickelt, welches standardmäßig utf8-Input-fähig war und moderne Schriftartkodierung unterstützt. Dann kam lualatex als man eine bessere Script-/Programmier-Sprache in TeX haben wollte als TeX selbst (*weil zu umständlich*), auch hier kann man Systemschriftarten nutzen.

Da  wir in dem Skript Liberation Sans als Schriftart setzen, von der ich mir fast absolut sicher bin, dass sie nicht in texlive ist, machen wir uns das leben erheblich leichter damit xelatex (oder lualatex) zu benutzen (und mithilfe von `fontspec` Liberation Sans zu laden) als uns damit zu quälen es [per Hand zu installieren](http://tug.ctan.org/info/Type1fonts/fontinstallationguide/fontinstallationguide.pdf).

Mister00X commented:
Nochmal was zu `hmargin` und `vmargin`, die funktionieren nicht, weil sie in dem Script an der falschen stelle sind (glaube ich). Im Pandoc Handbuch [steht](https://pandoc.org/MANUAL.html#general-writer-options) zu `-V key=val`:
> Set the template variable KEY to the value VAL when rendering the document in standalone mode. If no VAL is specified, the key will be given the value true.

Allerdings setzt pandoc für latex [gar keine `vmargin` oder `hmargin` variable](https://pandoc.org/MANUAL.html#variables-for-latex) in seinen Templates sondern eine variable bzw. einen YAML-Block für geometry. Wenn ich mir also https://github.com/siduction/sidu-manual/blob/6b6c4c15f10750cdf8cc10092a4ff9bded8537cf/development/10-generate-pdf-manual.sh#L122 anschaue, und das funktioniert, dann denke ich, dass man https://github.com/siduction/sidu-manual/blob/6b6c4c15f10750cdf8cc10092a4ff9bded8537cf/development/10-generate-pdf-manual.sh#L119-L120 wahrscheinlich zu
```sh
-V geometry:vmargin=$vmargin \
-V geometry:hmargin=$hmargin \
```
umändern muss, ob das so stimmt weiß ich nicht (gibt auch nix in der Dokumentation dazu), aber das könnte ich morgen testen, indem ich mir anschaue, ob der generierte LaTeX-Code die Optionen enthält.

P.S.: Apropos, ich hab gerade noch https://github.com/siduction/sidu-manual/blob/6b6c4c15f10750cdf8cc10092a4ff9bded8537cf/development/10-generate-pdf-manual.sh#L126 entdeckt und bin nun vollends verwirrt, welche Dinge jetzt am Ende tatsächlich im LaTeX-Code stehen. Allerdings haben wir in `11-...` auch noch ein `\usepackage{geometry}` stehen, welches damit zwei mal im Code auftaucht (das will ich in #52 entfernen), weshalb ich wirklich nicht sicher bin was der TeX-Compiler am Ende damit macht.

ak-li commented:
Entschuldige bitte. Ich habe den falschen Codeschnipse kopiert und damit genau das eingebaut was ich verhindern wollte.

In der Zeile

~~~
   if [ $# -eq 2 ] && [ $2 = --tex-debug ]
~~~

muss `$2` zu `'$2'` geändert werden, sonst wird Schadcode erst ausgeführt und dann findet die Überprüfung statt. Die einfachen Anführungszeichen weisen die bash an, den Inhalt nicht zu interpretieren.

Beispiel:
Die Eingabe von `./10-generate-pdf-manual.sh en $(echo --tex-debug)` führt ohne einfache Anführungszeichen zu

~~~
Creating LaTeX file instead of pdf.
~~~

und mit einfachen Anführungszeichen zu

~~~
More than two options or the second option is not '--tex-debug'
~~~

Im ersten Fall wird `$(echo --tex-debug)` ausgeführt und `--tex-debug` an den Test übergeben. Das ist bei Schadcode tödlich.


Das Gleiche gilt für die Zeilen

~~~
if [ "$1" = "de" ]
   und
if [ "$1" = "en" ]
~~~

Hier bei `$1` die doppelten Anführungszeichen durch einfache ersetzen. Bei `de` und `en` sind keine Anführungszeichen notwendig (aber sie schaden auch nicht), da der String keine Leerzeichen enthält.

ak-li commented:
> weshalb ich wirklich nicht sicher bin was der TeX-Compiler am Ende damit macht.

Konfusion komplett?
Ich werde mir noch einmal Zeit nehmen und deine Hinweise ausprobieren. Und anschließend bin ich dafür, alles was wir nicht benötigen aus den Skripten zu entfernen.

Darauf gekommen bin ich, weil mir der untere Seitenrand sehr groß erscheint. Unten ~37mm oben ~21mm.

Mister00X commented:
> Darauf gekommen bin ich, weil mir der untere Seitenrand sehr groß erscheint. Unten ~37mm oben ~21mm.

Das liegt meines Wissens nach an bestimmten typographischen Richtlinien an denen sich LaTeX orientiert. Unten wird in der Regel mehr Platz gelassen, weil die Seitenzahl nicht zum eigentlichen Text gehört und deshalb quasi unten zusätzlich dafür Platz gelassen wird.



> Und anschließend bin ich dafür, alles was wir nicht benötigen aus den Skripten zu entfernen.

In dem Fall möchte ich mal die Idee in den Raum werfen, die voreingestellten Variablen (zumindest einige davon) nicht mit `-V key=val` zu übergeben, sondern eine YAML Datei zu nutzen. Pandoc hat verschiedene Möglichkeiten Einstellungen aus YAML-Dateien zu übernehmen, eine davon ist das [*defaults file*](https://pandoc.org/MANUAL.html#defaults-files) eine andere [*metadata files*](https://pandoc.org/MANUAL.html#reader-options). Ich hab mir das selber noch nicht genauer angeschaut, dachte mir allerdings schon ein paar mal, dass das nützlich sein könnte.


Apropos ich habe [herrausgefunden](https://pandoc.org/demos.html), dass 
```sh
-V "geometry=vmargin=$vmargin" \
-V "geometry=hmargin=$hmargin" \
```
(auch) akzeptiert wird.

Mister00X commented:
Ach ja eine Sache noch gibt es einen Grund warum hier
```sh
    if [ $# -eq 2 ] && [ $2 = --tex-debug ]
```
`--tex-debug` nicht in Anführungszeichen sein muss/soll, ist das egal?

ak-li commented:
> warum hier `--tex-debug` nicht in Anführungszeichen sein muss/soll, ist das egal?

Nicht nötig, keine Metazeichen, keine Leerzeichen.

------

Habe das gefunden: https://de.overleaf.com/learn/latex/Page_size_and_margins.
Diese Parameter und Werte habe ich übernommen

~~~
textheight=227mm     (en 211mm)
voffset=6mm
     -V geometry:textheight=$textheight \
     -V geometry:voffset=$voffset \
~~~

In der debug-Datei wird daraus in Zeile 56 `\usepackage[textheight=231mm,voffset=6mm]{geometry}`.
Gleichzeitig habe ich aus `10-generatepdf-manual.sh`

~~~
#alignment=flushleft
#alignment=flushright
#alignment=center
#hmargin=2cm
#vmargin=2.0cm
#margin=20mm

#     -V alignment=$alignment \
#     -V geometry:margin="$margin" \
#     -V geometry:hmargin=$hmargin \
#     -V geomerty:vmargin=$vmargin \
#     -V geometry=$geometry \
~~~

entfernt und aus `11-helpfile-pdf-manual-header.tex`

~~~
\usepackage{geometry}
~~~

entfernt.
Das ergebnis ist für das de PDF 
Seitenrand oben 26 mm, bis zur horizontalen Linie 31 mm
Seitenrand unten 27 mm (unter der Seitenzahl), von unten bis zur horizontalen Linie 32 mm

Das sieht gut aus und wir haben unten nicht so einen dicken weißen Block aber dafür zwei Zeilen mehr Text. Die Seitenränder sind unverändert links 31 mm und rechts 30 mm. 

----------------
> In dem Fall möchte ich mal die Idee in den Raum werfen, die voreingestellten Variablen (zumindest einige davon) nicht mit `-V key=val` zu übergeben, sondern eine YAML Datei zu nutzen.

Ich sehe darin jetzt keinen Vorteil. Wir sollten lieber das aufräumen was wir haben.

Mister00X commented:
> > warum hier `--tex-debug` nicht in Anführungszeichen sein muss/soll, ist das egal?
> 
> Nicht nötig, keine Metazeichen, keine Leerzeichen.

Verstehe. Danke für die Erklärung! (Wieder was gelernt :wink:) 
In dem Fall werde ich nachher die PR entsprechend abändern.

> aus `11-helpfile-pdf-manual-header.tex`
> 
> ```
> \usepackage{geometry}
> ```
> 
> entfernt.

Gut! Das wollte ich in #52 eh machen.

> Das ergebnis ist für das de PDF Seitenrand oben 26 mm, bis zur horizontalen Linie 31 mm Seitenrand unten 27 mm (unter der Seitenzahl), von unten bis zur horizontalen Linie 32 mm
> 
> Das sieht gut aus und wir haben unten nicht so einen dicken weißen Block aber dafür zwei Zeilen mehr Text. Die Seitenränder sind unverändert links 31 mm und rechts 30 mm.

Alles klar, das klingt vernünftig.


> > In dem Fall möchte ich mal die Idee in den Raum werfen, die voreingestellten Variablen (zumindest einige davon) nicht mit `-V key=val` zu übergeben, sondern eine YAML Datei zu nutzen.
> 
> Ich sehe darin jetzt keinen Vorteil. Wir sollten lieber das aufräumen was wir haben.

Okay :+1:

Mister00X commented:
@ak-li Ich glaube jetzt müsste alles passen :grin:
